flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionMemory.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+
+import javax.annotation.Nullable;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Putting and getting of buffers to/from a region of memory.
+ * This class handles the headers, length encoding, memory slicing.
+ */
+final class BoundedBlockingSubpartitionMemory {
+
+	// all fields and methods below here have package-private access to avoid bridge
+	// methods when accessing them from the nested classes
+
+	static final int HEADER_LENGTH = 8;
+
+	static final int HEADER_VALUE_IS_BUFFER = 0;
+
+	static final int HEADER_VALUE_IS_EVENT = 1;
+
+	static ByteBuffer checkAndConfigureByteBuffer(ByteBuffer buffer) {
+		checkArgument(buffer.position() == 0);
+		checkArgument(buffer.capacity() > 8);
+		checkArgument(buffer.limit() == buffer.capacity());
+
+		return buffer.order(ByteOrder.nativeOrder());
+	}
+
+	// ------------------------------------------------------------------------
+
+	static final class Writer {
+
+		private final ByteBuffer memory;
+
+		Writer(ByteBuffer memory) {
+			this.memory = checkAndConfigureByteBuffer(memory);
+		}
+
+		public boolean writeBuffer(Buffer buffer) {
+			final ByteBuffer memory = this.memory;
+			final int bufferSize = buffer.getSize();
+
+			if (memory.remaining() < bufferSize + HEADER_LENGTH) {
+				return false;
+			}
+
+			memory.putInt(buffer.isBuffer() ? HEADER_VALUE_IS_BUFFER : HEADER_VALUE_IS_EVENT);
+			memory.putInt(bufferSize);
```

Comment:
these two `int` are written as metadata at the front of memory have total lenght `4+4=8` denoted by `HEADER_LENGTH `, am I right?

Reply:
yes, that is correct